### PR TITLE
rgw: refresh the generation of the bucket shard when fetching info

### PR DESCRIPTION
**Pull request by yuvalif**
when RGWRados::block_while_resharding() fails because reshard is in
progress, in the next iteration we should fetch the bucket shard
generation. for the case that the generation changed in the middle.

Signed-off-by: Yuval Lifshitz <ylifshit@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by cbodley — APPROVED** [review]
looks good! i just wish we had test cases that could reproduce/verify things like this

**Comment by cbodley** [comment]
i opened https://tracker.ceph.com/issues/54260 as a reminder to unify RGWRados::guard_reshard() and RGWRados::Bucket::UpdateIndex::guard_reshard()

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -6854,9 +6855,8 @@ int RGWRados::block_while_resharding(RGWRados::BucketShard *bs,
   // new_bucket_id and returns 0, otherwise it returns a negative
   // error code
   auto fetch_new_bucket_info =
-    [this, bs, &bucket_info, &bucket_attrs, &y, dpp](const std::string& log_tag) -> int {
-      int ret = get_bucket_info(&svc, bs->bucket.tenant, bs->bucket.name,
-                                bucket_info, nullptr, y, dpp, &bucket_attrs);
+    [this, bs, &obj_instance, &bucket_info, &bucket_attrs, &y, dpp](const std::string& log_tag) -> int {
+      int ret = bs->init(bs->bucket, obj_instance, &bucket_info, dpp);
```
[the remark this answers is not in the document]
Comment (yuvalif) [reply src/rgw/rgw_rados.cc:6859]: when i try to run that test i keep getting `radosgw-admin` crashes:
```
ceph version 17.0.0-10768-g929f175b54d (929f175b54d598b7fd7b63c6de1b86062b64c8dd) quincy (dev)                                                                                                                                               
 1: /lib64/libpthread.so.0(+0x13a20) [0x7f139eaf9a20]                                                                                                                                                                                         
 2: gsignal()                                                                                                                                                                                                                                 
 3: abort()                                                                                                                                                                                                                                   
 4: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x263) [0x7f139e1771ba]                                                                                                                                             
 5: (ceph::register_assert_context(ceph::common::CephContext*)+0) [0x7f139e17727a]                                                                                                                                                            
 6: (std::__detail::__variant::__gen_vtable_impl<std::__detail::__variant::_Multi_array<std::__detail::__variant::__deduce_visit_result<int> (*)(FaultInjector<std::basic_string_view<char, std::char_traits<char> > >::check(std::basic_strin
g_view<char, std::char_traits<char> > const&) const::visitor&&, std::variant<std::monostate, InjectAbort, InjectError> const&)>, std::integer_sequence<unsigned long, 1ul> >::__visit_invoke(FaultInjector<std::basic_string_view<char, std::c
har_traits<char> > >::check(std::basic_string_view<char, std::char_traits<char> > const&) const::visitor&&, std::variant<std::monostate, InjectAbort, InjectError> const&)+0xdf) [0x563cab507873]
 7: radosgw-admin(+0x8207be) [0x563cab5017be]
 8: radosgw-admin(+0x820b5e) [0x563cab501b5e]
 9: (RGWBucketReshard::execute(int, FaultInjector<std::basic_string_view<char, std::char_traits<char> > > const&, int, DoutPrefixProvider const*, bool, std::ostream*, ceph::Formatter*, RGWReshard*)+0xe1) [0x563cab5050af]
 10: main()    
 11: __libc_start_main()
 12: _start()
```

[the remark this answers is not in the document]
Comment (cbodley) [reply src/rgw/rgw_rados.cc:6859]: yes, the test uses fault injection to test crash recovery

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
thanks, tests are all passing again. merged

**Pull request closed without merging**